documentation/Data-View-Filters.md

```diff
@@ -0,0 +1,385 @@
+
+For example, if you have an "Address" column, that is build by combining a number of different fields (e.g. `address`, `province`, `country`), then you could specify any one of these as the value to sort by when clicking your custom "Address" column.
+
+It's also possible to convert complex data types to simplified "text" (or other simple data) versions of the values (see [Handling complex data structures](#handling-complex-data-structures) for more info), in which case you can specify this simplified column as the sort column for the more complex display value.
+
+### API
+
+Sorting requests are passed to the server as part of the query parameters of the `data-views/<table>` endpoint, e.g.
+
+```
+/data-views/product?orderBy=name&ascending=false
+```
+
+## Filters
+
+In the front end, there are several diffent UI elements for filtering table data. Which one is shown for a given data value is determined by a combination of the data type and explicit filter definitions specified in `data_view_column_definition`.
+
+### Handling complex data structures
+
+One of the main challenges with trying to create filters is that much of the data stored in tables is not easily filterable. Many fields store the entire "response" object that came from an application question, which could be any complex shape. For example, a `genericNames` field that has captured responses from a "Search" element might have data stored like so:
+
+```
+{
+  "text": "Ascorbic acid, Agni casti fructus, Alfuzosin and finasteride",
+  "selection": [
+    {
+      "name": "Ascorbic acid",
+      "atcCode": "G01AD03",
+      "category": "Gynecological antiinfectives and antiseptics"
+    },
+    {
+      "name": "Agni casti fructus",
+      "atcCode": "G02CX03",
+      "category": "Other gynecologicals"
+    },
+    {
+      "name": "Alfuzosin and finasteride",
+      "atcCode": "G04CA51",
+      "category": "Urologicals"
+    }
+  ]
+}
+```
+
+Clearly this is not filterable in any meaningful way with simple database queries. So we've provided a mechanism by which complex data values can be converted into more easily parse-able objects, such as plain text or numbers.
+
+We can define "filter data" columns, which are defined like so:
+
+- an extra column definition in `data_view_column_definitions`. It's just the same as other definitions, but the `column_name` field should have the suffix `FilterData` to be recognised by the system as a filter data column.
+- there are two fields in `data_view_column_definitions` that are of relevance:
+  - `filter_expression`: an [evaluator](Query-Syntax.md) expression defining the value for the filter data column. It's essentially the same as the [`value_expression` column](Data-View.md#data_view_column_definition-table) in that it takes the current record as the evaluator "objects" parameter.
```

Comment:
The link to `data_view_column_definition-table` seems broken

Reply:
At least I couldn't relate to the place that this link took me. And on Data-View.md the TOC is not taking anywhere in that page for these two subtitles:
![Screen Shot 2022-11-03 at 12 34 30 PM](https://user-images.githubusercontent.com/16461988/199621810-8fe42222-2c7a-47d1-aab1-62f7fdbcc14f.png)

Reply:
Seems to be working okay for me. Where are you viewing it?

If you look as a Github doc, it seems to work: https://github.com/openmsupply/conforma-server/blob/1389B-filter-tweaks/documentation/Data-View-Filters.md

(And those two TOC links seem okay too: https://github.com/openmsupply/conforma-server/blob/1389B-filter-tweaks/documentation/Data-View.md)

Reply:
I think this is not part of your PR, so don't feel like you have to fix it. But on the page you just linked: https://github.com/openmsupply/conforma-server/blob/1389B-filter-tweaks/documentation/Data-View.md

If you click those subheaders it doesn't take you anywhere, try cicking on the next one to see how it actually takes you to the Subheader selected.

Reply:
Ah, gotcha. Weird that the TOC links weren't auto-updated. Anyway, done that now: 1bfeaee9

(I think the one on the Filter page should be fine now too, if it wasn't already)